File: Programmers/42861.py
from collections import deque
def solution(n, costs):
    answer = 0
    costs.sort(key=lambda x : x[2])
    queue = []
    cost = 0
    for i in costs:
        queue.append(i)
    visited = deque()
    visited.append(0)
    now_d = 0
    while queue and len(visited) < n:
        if queue[now_d][0] in visited and queue[now_d][1] in visited:
            queue.pop(now_d)
            now_d = 0
        elif queue[now_d][0] in visited and queue[now_d][1] not in visited:
            cost += queue[now_d][2]
            visited.append(queue[now_d][1])
            queue.pop(now_d)
            now_d = 0
        elif queue[now_d][1] in visited and queue[now_d][0] not in visited:
            cost += queue[now_d][2]
            visited.append(queue[now_d][0])
            queue.pop(now_d)
            now_d = 0
        else:
            now_d += 1
    answer = cost
    return answer

#프로그래머스 레벨3 섬 연결하기
#그리디

File: Programmers/java/42861.java
import java.util.*;
class Solution {
    private int[] parent;
    public int find(int a) {
        if(parent[a] == a) return a;
        else return parent[a] = find(parent[a]);
    }
    
    public void union(int a, int b) {
        a = find(a);
        b = find(b);
        if(a != b) {
            parent[b] = a;
        }
    }
    
    public int solution(int n, int[][] costs) {
        int answer = 0;
        parent = new int[n];
        for(int i = 0; i < n; i++) {
            parent[i] = i;
        }
        Arrays.sort(costs, (o1, o2) -> o1[2] - o2[2]);
        //Kruskal Algorithm
        for(int i = 0; i < costs.length; i++) {
            if(find(costs[i][0]) != find(costs[i][1])) {
                union(costs[i][0], costs[i][1]);
                answer += costs[i][2];
            }
        }
        return answer;
    }
}

import java.util.*;
class Solution {
    
    public class Point implements Comparable<Point> {
        int node, cost;
        
        public Point(int node, int cost) {
            this.node = node;
            this.cost = cost;
        }
        
        @Override
        public int compareTo(Point o) {
            return this.cost - o.cost;
        }
    }
    
    public List<List<Point>> map = new ArrayList<>();
    
    public int solution(int n, int[][] costs) {
        int answer = 0;
        for(int i = 0; i < n; i++) {
            map.add(new ArrayList<>());
        }
        for (int i = 0; i < costs.length; i++) {
            int from = costs[i][0];
            int to = costs[i][1];
            int val = costs[i][2];
            map.get(from).add(new Point(to, val));
            map.get(to).add(new Point(from, val));
        }
        //Prim Algorithm
        boolean[] visited = new boolean[n];
        PriorityQueue<Point> queue = new PriorityQueue<>();
        queue.add(new Point(0, 0));
        while(!queue.isEmpty()) {
            Point cur = queue.poll();
            
            if(visited[cur.node]) continue;
            visited[cur.node] = true;
            answer += cur.cost;
            
            for(int i = 0; i < map.get(cur.node).size(); i++) {
                int next = map.get(cur.node).get(i).node;
                int cost = map.get(cur.node).get(i).cost;
                if(visited[next]) continue;
                queue.add(new Point(next, cost));
            }
        }
        return answer;
    }
}